File: src/java/WaveView.java
import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Build;
import android.os.CountDownTimer;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import java.util.ArrayList;

public class WaveView extends View {

    private int mWaveNum = 7;
    private float mRadiusMin = 90f;
    private float mRadiusMax = 1080f;
    private float mWaveInterval = 240f;
    private int mWaveColor = Color.WHITE;
    private int mWaveAlpha = 128;
    private int mBgColor = Color.TRANSPARENT;

    private Paint mPaint;
    private ArrayList<Float> mWaveList;
    private CountDownTimer waveAnim = new CountDownTimer(30000, 16) {
        @Override
        public void onTick(long millisUntilFinished) {
            for (int i = 0; i < mWaveList.size(); i++) {
                mWaveList.set(i, mWaveList.get(i) + 20f);
                if (mWaveList.get(i) < mRadiusMin + mWaveInterval){
                    mWaveList.set(i+1, mRadiusMin);
                    break;
                }
            }
            if (mWaveList.get(0) > mRadiusMax){
                mWaveList.set(0, 0f);
                ArrayList<Float> tempList = transList(mWaveList);
                mWaveList.clear();
                mWaveList.addAll(tempList);
            }

            invalidate();
        }

        @Override
        public void onFinish() {

        }
    };


    public WaveView(Context context) {
        super(context);
        init();
    }

    public WaveView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public WaveView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public WaveView(Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }


    private void init() {
        mPaint = new Paint();
        mWaveList = new ArrayList<>(mWaveNum);
        mPaint.setColor(mWaveColor);
        initRadius();
    }

    private void initRadius() {
        mWaveList.add(mRadiusMin);
        for (int i = 1; i < mWaveNum; i++) {
            mWaveList.add(0f);
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        float centerX = canvas.getWidth()/2;
        float centerY = canvas.getHeight()/2;
        canvas.drawColor(mBgColor);

        for (int i = 0; i < mWaveList.size(); i++) {
            mPaint.setAlpha(calcAlpha(mWaveList.get(i)));
            canvas.drawCircle(centerX, centerY, mWaveList.get(i), mPaint);
        }
    }


    private ArrayList<Float> transList(ArrayList<Float> list) {
        ArrayList<Float> newList = new ArrayList<>(mWaveNum);
        for (int i = 1; i < list.size(); i++) {
            newList.add(list.get(i));
        }
        newList.add(list.get(0));

        return newList;
    }

    private int calcAlpha(float radius){
        return (int)((mRadiusMax - radius)/mRadiusMax * mWaveAlpha);
    }

    public void startWave(){
        waveAnim.start();
    }

    public void stopWave(){
        waveAnim.cancel();
    }

    public void resetWave(){
        mWaveList.clear();
        mPaint.setAlpha(mWaveAlpha);
        mPaint.setColor(mWaveColor);
        initRadius();
        invalidate();
    }

    public int getmWaveNum() {
        return mWaveNum;
    }

    public void setmWaveNum(int mWaveNum) {
        this.mWaveNum = mWaveNum;
    }

    public float getmRadiusMin() {
        return mRadiusMin;
    }

    public void setmRadiusMin(float mRadiusMin) {
        this.mRadiusMin = mRadiusMin;
        mWaveList.clear();
        initRadius();
    }

    public float getmRadiusMax() {
        return mRadiusMax;
    }

    public void setmRadiusMax(float mRadiusMax) {
        this.mRadiusMax = mRadiusMax;
    }

    public int getmWaveColor() {
        return mWaveColor;
    }

    public void setmWaveColor(int mWaveColor) {
        this.mWaveColor = mWaveColor;
        mPaint.setColor(mWaveColor);
    }
}

File: src/kotlin/WaveView.kt
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.os.CountDownTimer
import android.util.AttributeSet
import android.view.View


class WaveView@JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {


    private val paint:Paint = Paint()
    public var mBgColor = Color.TRANSPARENT
    public var mWaveColor = Color.WHITE

    public var mRadiusMin = 90f
    public var mRadiusMax = 1080f
    public var mWaveInterval = 240f
    //用每次扩散半径增加的值作为速度参数
    public var speed = 14f

    private val mWaveList = ArrayList<Float>()

    private val timeline = object:CountDownTimer(300000,16){

        override fun onTick(millisUntilFinished: Long) {

            // 每个时间间隔都把正在扩散的波纹半径增加
            for (i in 0 until mWaveList.size){
                mWaveList[i] = mWaveList[i] + speed
                if (mWaveList[i] < mRadiusMin + mWaveInterval){
                    break
                }
            }

            //最外层波纹超过最大值时，重新把它添加到波纹队列末尾
            if (mWaveList[0] > mRadiusMax){
                mWaveList[0] = mRadiusMin
                val newList = transList(mWaveList)
                mWaveList.clear()
                mWaveList.addAll(newList)
            }

            invalidate()
        }

        override fun onFinish() {
            //尽量保证手动调用waveStop，就不会执行到这里
            reset()
        }
    }

    init {
        initWave(mWaveList)
        paint.color = mWaveColor
    }


    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        canvas!!.drawColor(mBgColor)
        val centerX = canvas.width.div(2).toFloat()
        val centerY = canvas.height.div(2).toFloat()

        for (i in 0 until mWaveList.size){
            paint.alpha = calcAlpha(mWaveList[i])
            canvas.drawCircle(centerX, centerY, mWaveList[i], paint)
        }
    }

    fun wave(){
        timeline.start()
    }

    fun stopWave(){
        timeline.cancel()
    }

    fun reset(){
        timeline.cancel()
        mWaveList.clear()
        initWave(mWaveList)
    }

    private fun initWave(waveList: ArrayList<Float>){
        val waveNum = ((mRadiusMax-mRadiusMin)/mWaveInterval).toInt() + 2
        for (i in 1..waveNum){
            waveList.add(mRadiusMin)
        }
    }

    private fun transList(list: ArrayList<Float>):ArrayList<Float>{
        val newList = ArrayList<Float>()

        (1 until list.size).mapTo(newList) { list[it] }
        newList.add(list[0])
        return newList
    }

    // 通过半径计算透明度，趋势是半径越大越透明，直到看不见
    private fun calcAlpha(r:Float):Int = ((mRadiusMax - r)/ mRadiusMax * 120).toInt()
}